### ISSUE TITLE
Play 2.3.x not detecting changes to static resources in shared folder

### ISSUE THREAD

**Issue by johnhoffman** [playframework/playframework#4208]
The solution that @jroper provided in https://github.com/playframework/playframework/issues/3246 works, but from my experience it doesn't work with static assets such as HTML and JavaScript files that are not templates (doesn't matter if they're in the public folder or the app/assets folder). Rather, one has to stop, compile, and run again manually to get the changes to take effect. As an additional pain point, this unnecessarily kills sessions. It can take a pretty big toll on productivity.

If it matters, I'm also running Activator in a Docker container.

Is there a solution for this that I'm missing?

**Comment by johnhoffman** [comment]
To be a little more specific, if I save a change on a static resource (i.e, foobar.html) on my dev machine, I can see that the update syncs all the way to the docker container. However, when I refresh the page, the server sends the old static resource with a 200 response. So it appears that static resources changes are not detected, and the static resources in the running application are therefore not refreshed. When do this on my dev machine (not a shared folder), the static file is updated w.r.t the running application without a recompile.

**Comment by jroper** [comment]
When the sync is done, is the modification time of the file being updated too?

**Comment by johnhoffman** [comment]
I think so, but not as I'd expect. This is before making a change to main.html:

    [vagrant@vagrant assessments]$ ls -l
    total 12
    -rw-r--r--. 1 503 games 1881 Apr  8 18:36 controllers.js
    -rw-r--r--. 1 503 games 3147 Apr  9 12:39 main.html
    -rw-r--r--. 1 503 games  232 Apr  7 18:10 main.js

This is after making a change:

    [vagrant@vagrant assessments]$ ls -l
    total 12
    -rw-r--r--. 1 503 games 1881 Apr  8 18:36 controllers.js
    -rw-r--r--. 1 503 games 3147 Apr  9  2015 main.html
    -rw-r--r--. 1 503 games  232 Apr  7 18:10 main.js

Is there a better way to check this?

**Comment by jroper** [comment]
We could be on to something here.  Try using the `stat` command instead, on Linux:

    $ stat -t main main.html

This should show the raw timestamp values.  I'm not sure, but it looks like the timestamp on `main.html` is losing precision after an update, to one day, which means the sbt change detection algorithm, which is based on the timestamp being more recent then when it was last checked, would not work.

**Comment by johnhoffman** [comment]
The timestamp was being updated, usually with a significant lag (~10 seconds), whereas the file contents were updated immediately. So now I've made a change my Vagrantfile from this:

    config.vm.synced_folder ".", "/vagrant", type: "nfs", nfs_export: false, nfs_udp: false

to this:

    config.vm.synced_folder ".", "/vagrant", type: "nfs", nfs_export: false, nfs_udp: false, mount_options: ["lookupcache=none"]

Now the timestamp is updated immediately (and without the weird ```ls -l``` formatting quirk).

Before modifying ```main.html```:

    [root@883b8c551a75 assessments]# stat main.html
      File: `main.html'
      Size: 3147      	Blocks: 8          IO Block: 65536  regular file
    Device: 28h/40d	Inode: 3294893     Links: 1
    Access: (0644/-rw-r--r--)  Uid: (  503/ UNKNOWN)   Gid: (   20/   games)
    Access: 2015-04-10 15:25:31.000000000 +0000
    Modify: 2015-04-10 15:29:25.000000000 +0000
    Change: 2015-04-10 15:29:25.000000000 +0000

After modifying ```main.html```:

    [root@883b8c551a75 assessments]# stat main.html
      File: `main.html'
      Size: 3148      	Blocks: 8          IO Block: 65536  regular file
    Device: 28h/40d	Inode: 3294893     Links: 1
    Access: (0644/-rw-r--r--)  Uid: (  503/ UNKNOWN)   Gid: (   20/   games)
    Access: 2015-04-10 15:25:31.000000000 +0000
    Modify: 2015-04-10 15:29:41.000000000 +0000
    Change: 2015-04-10 15:29:41.000000000 +0000

You can see that the seconds changed from 25 to 41 as expected. This looks like progress.

Unfortunately though, refreshing the page in the browser still doesn't trigger an update to the static content in the running application (the server returns the unchanged file). This line is in my ```build.sbt```:

    PlayKeys.playWatchService := play.sbtplugin.run.PlayWatchService.sbt(pollInterval.value)

**Comment by shkhln** [comment]
Oh, I've seen that too. (But haven't bothered to look at issue closely until today.) The `.distinct` [here]( https://github.com/playframework/playframework/blob/ecc1a1f175eb85b6d89ac893d75973c612cca942/framework/src/run-support/src/main/scala/play/runsupport/FileWatchService.scala#L119) operates only on file names, not full paths, which means something like `views/a/main.html, views/b/main.html, views/c/main.html, views/other.html` becomes `views/a/main.html, views/other.html`.

**Comment by johnhoffman** [comment]
The ```Reload triggered``` message did not print to the sbt console when I modified a static resource, even after several page refreshes. It did print when I made a change to ```app/controllers/Application.scala``` and then refreshed the page a few times.

In light of @shkhln's comment, I tried modifying a static resource that has a unique filename. Unfortunately, this didn't trigger a reload either.

**Comment by shkhln** [comment]
That's odd. There can be multiple issues of course. Can you please check that again with the following config.

build.sbt:
```
PlayKeys.playWatchService := new MySbtWatchService(pollInterval.value)
```

project/MySbtWatchService.scala:
```
import play.sbtplugin.run.PlayWatchService
import play.sbtplugin.run.PlayWatcher

import java.io.File

import sbt.{ SourceModificationWatch, WatchState }
import sbt.{ IO, DirectoryFilter, HiddenFileFilter }
import sbt.Path._

class MySbtWatchService(val pollDelayMillis: Int) extends PlayWatchService {

  def watch(filesToWatch: Seq[File], onChange: () => Unit) = {

    @volatile var stopped = false

    val thread = new Thread(new Runnable {
      def run() = {
        println("filesToWatch:\n" + filesToWatch.***)
        var state = WatchState.empty
        while (!stopped) {
          val (triggered, newState) = SourceModificationWatch.watch(filesToWatch.***, pollDelayMillis,
            state)(stopped)
          if (triggered) onChange()
          state = newState
        }
      }
    }, "sbt-play-watch-service")
    thread.setDaemon(true)
    thread.start()

    new PlayWatcher {
      def stop() = stopped = true
    }
  }
}
```

**Comment by johnhoffman** [comment]
@shkhln your MySbtWatchService appears to have resolved the issue. I now get the ```Reload triggered``` message, and the server returns the modified asset.

**Comment by johnhoffman** [comment]
@shkhln I just confirmed that it works for both Vagrant boxes I'm using (one with devicemapper, the other with BTRFS). There is some lag, but I suppose that's due to the pollingInterval and the time it takes to complete the file scan. This is a major improvement.

**Comment by jroper** [comment]
So the only different between that and the built in Play one is the lack of a `distinct` call?  That's odd, because `distinct` uses `Object.equals` equality, and for `File`, that *should* be based on the full path.  What happens if you change the `println` statement to:

```scala
println("filesToWatch:\n" + filesToWatch.***.distinct)
```

Are there any files missing?

**Comment by shkhln** [comment]
Sorry, I always forget to mention something. It is the (rather poorly named) [sbt.PathFinder.distinct](https://github.com/sbt/sbt/blob/127081b3c5af07058a88f0130e0cfd7fe0838e7a/util/io/src/main/scala/sbt/Path.scala#L176) method, not stdlib one.

**Comment by jroper** [comment]
Well that explains it.

**Comment by jroper** [comment]
Fixed in #4236.
